cli/src/commands/rebase.rs

```diff
@@ -348,91 +351,107 @@ fn rebase_descendants_transaction(
     Ok(())
 }
 
-fn rebase_revision(
+fn rebase_revisions(
     ui: &mut Ui,
     settings: &UserSettings,
     workspace_command: &mut WorkspaceCommandHelper,
     new_parents: &[Commit],
-    rev_arg: &RevisionArg,
+    target_commits: Vec<Commit>,
 ) -> Result<(), CommandError> {
-    let to_rebase_commit = workspace_command.resolve_single_rev(rev_arg)?;
-    workspace_command.check_rewritable([to_rebase_commit.id()])?;
-    if new_parents.contains(&to_rebase_commit) {
-        return Err(user_error(format!(
-            "Cannot rebase {} onto itself",
-            short_commit_hash(to_rebase_commit.id()),
-        )));
+    if target_commits.is_empty() {
+        return Ok(());
+    }
+
+    workspace_command.check_rewritable(target_commits.iter().ids())?;
+    for commit in target_commits.iter() {
+        if new_parents.contains(commit) {
+            return Err(user_error(format!(
+                "Cannot rebase {} onto itself",
+                short_commit_hash(commit.id()),
+            )));
+        }
     }
 
     let mut tx = workspace_command.start_transaction();
-    let mut rebased_commit_ids = HashMap::new();
+    let tx_description = if target_commits.len() == 1 {
+        format!("rebase commit {}", target_commits[0].id().hex())
+    } else {
+        format!(
+            "rebase commit {} and {} more",
+            target_commits[0].id().hex(),
+            target_commits.len() - 1
+        )
+    };
+
+    // First, rebase the descendants of `target_commits`.
+    let (initial_rebased_commit_ids, target_roots) =
+        extract_commits(&mut tx, settings, &target_commits)?;
+
+    // We now update `new_parents` to account for the rebase of all of
+    // `target_commits`'s descendants. Even if some of the original `new_parents`
+    // were descendants of `target_commits`, this will no longer be the case after
+    // the update.
+    let new_parents = new_parents
+        .iter()
+        .map(|new_parent| {
+            get_possibly_rewritten_commit_id(&initial_rebased_commit_ids, new_parent.id())
+        })
+        .collect_vec();
```

Comment:
Can we use `tx.mut_repo().new_parents()` here and not need `initial_rebased_commit_ids`?

Reply:
Right, this is a bit of a holdover from the old code, but I also noticed this comment in `transform_descendants` which indicates that `parent_mappings` might be cleared in the future, so I stuck with this approach for now.

Reply:
Maybe we can get rid of `initial_rebased_commit_ids` by switching to `mut_repo().new_parents()`?

(I don't know if we'll change the contract of `transform_descendants()`, though.)

Reply:
Maybe we should confirm if the API will change first?

Reply:
My only motivation for changing the API is so a `transform_descendants()` followed by `rebase_descendants()` can avoid needlessly looking for commits to rebase. An alternative way of solving that is to add a dirty bit saying if new commits have been marked rewritten/abandoned in between the two calls. We'll also need to keep in mind that there can be further rebasing necessary right after `transform_descendants()` since the callback might have rewritten unrelated commits, though we can possibly decide that that's misuse of the API that we don't want to care about.

I think we should be able to rely on `new_parents()` for now. If we decide to solve the problem above by clearing the set, then we'll need to make sure that we provide the functionality in a different way (probably by returning the map). But I think it's more likely that we'll keep `new_parents()` working.

Reply:
I've updated the code to not use `new_parents()`, and remove the usage of the commit ID map.